### DISCUSSION TITLE
Is it possible to add watermark on print document using pappl?

### DISCUSSION THREAD

**Discussion by trinhtran-markany** [michaelrsweet/pappl#316] [General]
Hi,

I'm developing a macOS application to trigger any printing action and modify the document content before sending to the printer, such as adding text to printing documents.

I would like to ask if the development is possible on macOS using **pappl** or **cups**? Regards!

**Comment by michaelrsweet** [comment]
PAPPL isn't intended for use this way - it can be used to implement IPP print services that macOS can send jobs to but it isn't a filtering mechanism per se.

CUPS (as shipped with macOS) offers limited extension capabilities due to sandboxing. But the standard print dialog already offers watermarking...

**Reply by trinhtran-markany** [reply]
@michaelrsweet Is it possible to develop a generic filter using CUPS (macOS) that can be applied to any kind of printer, forcing the user to add a default watermark without using a watermarking print dialog option?

**Reply by michaelrsweet** [reply]
Conceptually yes, in that you can define PDF to xxx filters with a lower cost than the system-supplied ones and install the xxx.convs file in /etc/cups, however that method is hard to get right, can be overridden by individual drivers, and could break at any time. You also need to either duplicate the original filters' functions or call them to produce the proper output.

**Reply by trinhtran-markany** [reply]
@michaelrsweet  On macOS(Sonoma), I tried your suggestion above:

1. Update PPD file by adding `*cupsFilter2: application/pdf application/vnd.pdf-watermark 0 watermark` 
2. Create a new file named `xxx.convs`.
3. Add filter `application/pdf application/vnd.pdf-watermark 0 watermark` to `xxx.convs`.
5. Move the file `xxx.convs` to `/usr/share/cups/mine` however seems impossible due to permission issues.

Could you help me -
- Is it correct flow to insert my filter to filter chain?
- The error happens in step 4, do you have any other solutions or workarounds?

I apologize for the numerous questions, but it seems there are no other places to address my concerns!

**Reply by michaelrsweet** [reply]
You need to intercept the original filters (which you will find listed in the /usr/share/cups/mime/*.convs files) by providing your own xxx.convs file in /etc/cups with costs less than the originals. Trying to do an intermediate MIME media type won't work because CUPS will try to optimize the filter chain to the lowest cost and fewest filters.

You are correct that you can't install to /usr/share/cups - macOS has had a read-only root filesystem for many years now - but you *can* install to /etc/cups.

**Reply by trinhtran-markany** [reply]
Thank you a lot, @michaelrsweet, for following up on my question and your quick response. I appreciate it. Here's what I did so far -

1. Create a `watermark.convs` file with below info:
```
# Watermark
application/pdf application/vnd.pdf-watermark 0 watermark
```

2. Create a `watermark.types` file with below info:
```
# Watermark
application/vnd.pdf-watermark 
```

3. Move both files to `/etc/cups`
4. Change my `MyPrinter.pdd` file :

from
```
*cupsFilter: "application/pdf 0 /Library/Printers/Canon/CUPSCAPT2/Bins/capdftopdl"
```
to 
```
*cupsFilter: application/vnd.pdf-watermark 0 /Library/Printers/Canon/CUPSCAPT2/Bins/capdftopdl
```
5. Create a new `watermark` filter and move to `/usr/libexec/cups/filter`. The watermark filter should simply print the log for testing purposes.
```
# Create a temporary directory for the script
tempdir="$(mktemp -d)"

echo "[Watermark] $tempdir" >&2
# Check if the temporary directory was successfully created
if [ ! -d "$tempdir" ]; then
    echo "Error: Unable to create temporary directory." >&2
    exit 1
fi

# Define the log file within the temporary directory
logfile="$tempdir/watermarkpdf.log"
watermark="/etc/cups/watermark.pdf"

echo "[Watermark] $0 $* (tempdir: $tempdir)" >&2

echo "[Watermark] $0 $* (tempdir: $tempdir)" >> $logfile

# Command line arguments
job="$1"
user="$2"
title="$3"
numcopies="$4"
options="$5"
filename="$6"
output="$7"

echo "[Watermark] $0 - $1 - $2 - $3 - $4 - $5 - $6 - $7" >&2

# Exit successfully
exit 0
```
7. Check the error log
```
D [28/Dec/2023:11:24:02 +0700] [Job 6] 2 filters for job:
D [28/Dec/2023:11:24:02 +0700] [Job 6] watermark (application/pdf to application/vnd.pdf-watermark, cost 0)
D [28/Dec/2023:11:24:02 +0700] [Job 6] /Library/Printers/Canon/CUPSCAPT2/Bins/capdftopdl (application/vnd.pdf-watermark to printer/Canon_LBP3300, cost 0)
D [28/Dec/2023:11:24:02 +0700] [Job 6] job-sheets=none,none
...
I [28/Dec/2023:11:24:02 +0700] [Job 6] Started filter /usr/libexec/cups/filter/watermark (PID 6886)
I [28/Dec/2023:11:24:02 +0700] [Job 6] Started filter /Library/Printers/Canon/CUPSCAPT2/Bins/capdftopdl (PID 6887)
I [28/Dec/2023:11:24:02 +0700] [Job 6] Started backend /usr/libexec/cups/backend/cnbma2 (PID 6888)
...
D [28/Dec/2023:11:25:23 +0700] [CGI] cgiSetArray: job_id[1]=\"4\"
D [28/Dec/2023:11:25:23 +0700] [CGI] cgiSetArray: job_state[1]=\"6\"
D [28/Dec/2023:11:25:23 +0700] [CGI] cgiSetArray: job_impressions_completed[1]=\"0\"
D [28/Dec/2023:11:25:23 +0700] [CGI] cgiSetArray: job_k_octets[1]=\"29\"
D [28/Dec/2023:11:25:23 +0700] [CGI] cgiSetArray: job_hold_until[1]=\"no-hold\"
D [28/Dec/2023:11:25:23 +0700] [CGI] cgiSetArray: job_printer_state_message[1]=\"Filter failed\"
D [28/Dec/2023:11:25:23 +0700] [CGI] cgiSetVariable: SERVER_NAME=\"localhost\"
D [28/Dec/2023:11:25:23 +0700] [CGI] cgiSetVariable: CUPS_VERSION=\"CUPS v2.3.4\"
D [28/Dec/2023:11:25:23 +0700] [CGI] cgiSetArray: job_preserved[2]=\"1\"
D [28/Dec/2023:11:25:23 +0700] [CGI] cgiSetArray: job_printer_name[2]=\"Canon_LBP3300\"
D [28/Dec/2023:11:25:23 +0700] [CGI] cgiSetArray: job_printer_uri[2]=\"/printers/Canon_LBP3300\"
D [28/Dec/2023:11:25:23 +0700] [CGI] cgiSetArray: time_at_completed[2]=\"novalue\"
D [28/Dec/2023:11:25:23 +0700] [CGI] cgiSetArray: time_at_creation[2]=\"Thu Dec 28 11:04:33 2023\"
D [28/Dec/2023:11:25:23 +0700] [CGI] cgiSetArray: time_at_processing[2]=\"Thu Dec 28 11:04:33 2023\"
```

I encountered an error marked as **Filter failed**. Do you have any insights or ideas on how to resolve this issue?

**Comment by michaelrsweet** [comment]
Your watermark filter needs to write PDF to stdout, otherwise the downstream filter (capdfgopdl in this case) won't have anything to print.
